torch/csrc/jit/autodiff.cpp

```diff
@@ -75,24 +93,32 @@ static value_set findAllRequiresGradNodes(
   return requires_grad_set;
 }
 
-static Value* allocZerosLike(Value *v) {
-  static const Symbol constant_sym = "constant"_sym;
-  static const Symbol is_zero_sym  = "is_zero"_sym;
-  static const Symbol value_sym    = "value"_sym;
+static Value* createZerosLike(Value *v) {
   JIT_EXPECTM(v->hasType(), "can't allocate zero gradient for a value without a type");
   Graph *graph = v->owningGraph();
   auto type = v->type()->expect<TensorType>();
   AutoGPU gpu_guard(type->device());
 
   auto & at_type = type->device() == -1 ? at::CPU(type->scalarType()) : at::CUDA(type->scalarType());
   auto zeros = at_type.zeros({1}).expand(type->sizes());
-  Node *constant = graph->create(constant_sym)
-                        ->t_(value_sym, zeros)
-                        ->i_(is_zero_sym, 1);
+  Node *constant = graph->createConstant(zeros)
+                        ->i_(kis_zero, 1);
   graph->appendNode(constant);
   return constant->output();
 }
 
+// any vjp input may be undefined, and we need to potentially replace it
+// with a zero tensor of the right size if required.
+// this function inserts a guard into the graph that does this replacement.
+// ReplaceIfUndef(dv,c) replaces dv with c if dv is undef.
+// During Graph specialization these guards will get removed when
+// 'dv' is known to be undef, and the zeros will be propagated if possible.
+static Value* createUndefGuard(Value * dv, Value * alternative) {
+  Graph* graph = dv->owningGraph();
+  Node * n = graph->create(kReplaceIfUndef, {dv, alternative});
+  return graph->appendNode(n)->output();
+}
```

Comment:
It does not get removed in all cases. So it has to be in the graph to make the graph semantically correct. This is not "polluting" the graph. It is correctly specifying what should happen, and then cleaning up the trivial cases later using rewrite rules. This has to happen because the autograd and the autodiff handle Undefined tensors differently.